// @polkadot/typegen/cjs/generate/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "generateDefaultCalls", {
  enumerable: true,
  get: function () {
    return _runtime.generateDefaultCalls;
  }
});
Object.defineProperty(exports, "generateDefaultConsts", {
  enumerable: true,
  get: function () {
    return _consts.generateDefaultConsts;
  }
});
Object.defineProperty(exports, "generateDefaultErrors", {
  enumerable: true,
  get: function () {
    return _errors.generateDefaultErrors;
  }
});
Object.defineProperty(exports, "generateDefaultEvents", {
  enumerable: true,
  get: function () {
    return _events.generateDefaultEvents;
  }
});
Object.defineProperty(exports, "generateDefaultInterface", {
  enumerable: true,
  get: function () {
    return _interfaceRegistry.generateDefaultInterface;
  }
});
Object.defineProperty(exports, "generateDefaultLookup", {
  enumerable: true,
  get: function () {
    return _lookup.generateDefaultLookup;
  }
});
Object.defineProperty(exports, "generateDefaultQuery", {
  enumerable: true,
  get: function () {
    return _query.generateDefaultQuery;
  }
});
Object.defineProperty(exports, "generateDefaultRpc", {
  enumerable: true,
  get: function () {
    return _rpc.generateDefaultRpc;
  }
});
Object.defineProperty(exports, "generateDefaultTsDef", {
  enumerable: true,
  get: function () {
    return _tsDef.generateDefaultTsDef;
  }
});
Object.defineProperty(exports, "generateDefaultTx", {
  enumerable: true,
  get: function () {
    return _tx.generateDefaultTx;
  }
});
var _consts = require("./consts");
var _errors = require("./errors");
var _events = require("./events");
var _interfaceRegistry = require("./interfaceRegistry");
var _lookup = require("./lookup");
var _query = require("./query");
var _rpc = require("./rpc");
var _runtime = require("./runtime");
var _tsDef = require("./tsDef");
var _tx = require("./tx");
