// @astar-network/astar-api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.options = exports.defaultOptions = void 0;
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
const astar_api_derive_1 = require("@astar-network/astar-api-derive");
const astar_types_1 = require("@astar-network/astar-types");
const orml_api_derive_1 = require("@open-web3/orml-api-derive");
exports.defaultOptions = {
    types: astar_types_1.types,
    rpc: astar_types_1.rpc
};
const options = ({ types = {}, rpc = {}, typesAlias = {}, typesBundle = {}, ...otherOptions } = {}) => {
    var _a, _b, _c, _d, _e, _f;
    return ({
        types: {
            ...astar_types_1.types,
            ...types
        },
        rpc: {
            ...astar_types_1.rpc,
            ...rpc
        },
        typesAlias: {
            ...astar_types_1.typesAlias,
            ...typesAlias
        },
        derives: {
            ...orml_api_derive_1.derive,
            ...astar_api_derive_1.derive
        },
        typesBundle: {
            ...typesBundle,
            spec: {
                ...typesBundle.spec,
                astar: {
                    ...(_a = astar_types_1.typesBundle === null || astar_types_1.typesBundle === void 0 ? void 0 : astar_types_1.typesBundle.spec) === null || _a === void 0 ? void 0 : _a.astar,
                    ...(_b = typesBundle === null || typesBundle === void 0 ? void 0 : typesBundle.spec) === null || _b === void 0 ? void 0 : _b.astar
                },
                shiden: {
                    ...(_c = astar_types_1.typesBundle === null || astar_types_1.typesBundle === void 0 ? void 0 : astar_types_1.typesBundle.spec) === null || _c === void 0 ? void 0 : _c.shiden,
                    ...(_d = typesBundle === null || typesBundle === void 0 ? void 0 : typesBundle.spec) === null || _d === void 0 ? void 0 : _d.shiden
                },
                shibuya: {
                    ...(_e = astar_types_1.typesBundle === null || astar_types_1.typesBundle === void 0 ? void 0 : astar_types_1.typesBundle.spec) === null || _e === void 0 ? void 0 : _e.shibuya,
                    ...(_f = typesBundle === null || typesBundle === void 0 ? void 0 : typesBundle.spec) === null || _f === void 0 ? void 0 : _f.shibuya
                }
            }
        },
        ...otherOptions
    });
};
exports.options = options;
